File: android-native-shell-scaffold/app/src/main/java/com/vinesight/shell/OnboardingFlowViewModel.kt
package com.vinesight.shell

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class OnboardingFlowViewModel {
    private val _step = MutableStateFlow(OnboardingStepId.LANGUAGE)
    val step: StateFlow<OnboardingStepId> = _step.asStateFlow()

    private val _preferredLanguageCode = MutableStateFlow("en")
    val preferredLanguageCode: StateFlow<String> = _preferredLanguageCode.asStateFlow()

    private val _notificationsEnabled = MutableStateFlow(true)
    val notificationsEnabled: StateFlow<Boolean> = _notificationsEnabled.asStateFlow()

    fun updatePreferredLanguageCode(value: String) {
        _preferredLanguageCode.value = value
    }

    fun updateNotificationsEnabled(value: Boolean) {
        _notificationsEnabled.value = value
    }

    fun next() {
        dispatch(OnboardingFlowEvent.Next)
    }

    fun previous() {
        dispatch(OnboardingFlowEvent.Previous)
    }

    fun complete() {
        dispatch(OnboardingFlowEvent.Complete)
    }

    fun reset() {
        dispatch(OnboardingFlowEvent.Reset)
    }

    fun skipToComplete() {
        dispatch(OnboardingFlowEvent.SkipToComplete)
    }

    private fun dispatch(event: OnboardingFlowEvent) {
        _step.value = OnboardingFlowCoordinator.nextStep(_step.value, event)
    }
}

File: ios-native-shell-scaffold/VineSightShell/OnboardingFlowViewModel.swift
import Foundation
import SwiftUI

@MainActor
final class OnboardingFlowViewModel: ObservableObject {
    @Published private(set) var step: OnboardingStepId = .language
    @Published var preferredLanguageCode: String = "en"
    @Published var notificationsEnabled: Bool = true

    func next() {
        step = OnboardingFlowCoordinator.nextStep(from: step, event: .next)
    }

    func previous() {
        step = OnboardingFlowCoordinator.nextStep(from: step, event: .previous)
    }

    func complete() {
        step = OnboardingFlowCoordinator.nextStep(from: step, event: .complete)
    }

    func skipToComplete() {
        step = OnboardingFlowCoordinator.nextStep(from: step, event: .skipToComplete)
    }

    func reset() {
        step = OnboardingFlowCoordinator.nextStep(from: step, event: .reset)
    }
}
